### PR TITLE
`LocalVector` small internal changes

### PR DESCRIPTION
**Pull request by dimkKy**
A follow-up to [#87794](https://github.com/godotengine/godot/pull/87794).

`LocalVector` `if(tight)` checks are now compile-time, and `remove_at()` and `insert()` now use `std::memmove()` for _trivially copyable_ types to shift data.

Please take a look at lines **89** and **247**. I'm not sure if I understood `force_trivial` parameter correctly.

### CODE REVIEW

**Comment by clayjohn** [comment]
Could you clarify what the purpose of this change is? Does it improve performance and/or memory usage? If so, could you post benchmarks that show the relative improvement that this PR makes?

For context, changes to core are super high risk as they have the potential to break things anywhere in the engine in hard to debug ways. So we don't make changes to core unless we have a compelling reason

**Comment by RandomShaper** [comment]
I think that doing byte-wise copy of trivially-constructible types is good. The resulting code may or may not be faster than otherwise, but never slower.

Regarding, `force_trivial`, the idea is, "treat `T` as if it was a trivial type regardless it is." That said, ideally we'd check in which case that's used to be sure about the intent.

Now, when it comes to upgrading the checks based on template parameters, such as `tight`, to `if constexpr`, I'd suggest to hold on. If that's really beneficial somehow, we should consider following the same pattern throughout the codebase. A potential benefit is faster execution (although in a case like this I doubt the compiler couldn't already generate optimal code) and another one is higher explicitness, but, again, is debatable, given that template parameters should already stand out (if not, like here with `tight`, let's check the naming convention).

In short:
1. I'd suggest constraining this PR to the do-zombie-copies-if-trivial part, which I think is clearly good. And maybe apply the same enhancement to `CowData`.
2. Open a proposal about upgrading checks which are known to be compile-time to the `if constexpr` syntax, and see what happens.

**Comment by dimkKy** [comment]
I think proposed changes are quite trivial and self-explanatory: just a little refactoring with slightly better exception safety.

**Comment by Nazarwadim** [comment]
I made some benchmarks on `memmove` `std::copy` and 
`			for (U i = p_index; i < count - 1; i++) {
				data[i] = data[i + 1];
			}`
https://quick-bench.com
image
Benchmark code (arrays are of the same size):
```
#include <cstring>
#include <memory>
static void MemMove(benchmark::State& state) {
  for (auto _ : state) {
  int data_my[] = {5113,4,135,123,5213,251,2135,14,1234,1234,123,4,43134,513,5,1354,134,31,4,134,134,1324,123,41,5134,64,56,548,567867,84,56754,34,456,24,245,2345,345,3245};
  int count = sizeof(data_my)/sizeof(int);
    memmove(&data_my[0], &data_my[0 + 1], (count -1) * sizeof(int));
    benchmark::DoNotOptimize(data_my);
  }
}

BENCHMARK(MemMove);

static void StdCopy(benchmark::State& state) {  
  for (auto _ : state) {
  int data_my[] = {65113,64,135,61361,5613,651,635,164,1234,12324,123,411,43134,513,5,1354,134,31,4,134,134,1324,121233,41,5134,64,56,548,56712317,8242414,23456754,34,456,24,245,2345,345,3245};
  int count = sizeof(data_my)/sizeof(int);
    std::copy(data_my, data_my + count - 1, data_my + 1);
    benchmark::DoNotOptimize(data_my);
  }
}
BENCHMARK(StdCopy);

static void ForLoop(benchmark::State& state) {  
  for (auto _ : state) {
  int data_my[] = {3,4,5,1361,13,651,635,164,1234,1234,123,411,43134,513,5,1354,134,31,4,134,134,1324,1233,41,5134,64,56,548,5677,8242414,23456754,34,456,24,245,2345,345,3245};
  int count = sizeof(data_my)/sizeof(int);
    for (size_t i = 0; i < count - 1; i++) {
	data_my[i] = data_my[i + 1];
     }
    benchmark::DoNotOptimize(data_my);
  }
}
BENCHMARK(ForLoop);
```
As it turned out, the usual for loop is a bit faster. This is because of the compiler's optimisation, which optimises the loop.
`std::copy` isn't faster than `memmove`.
If you look at the memmove source code, you will notice that there is an unnecessary if here. Not sure, but there is also an unnecessary function call that takes up resources.
```
void *
memmove (void *dest, const void *src, size_t len)
{
  char *d = dest;
  const char *s = src;
  if (d < s)
    while (len--)
      *d++ = *s++;
  else
    {
      char *lasts = s + (len-1);
      char *lastd = d + (len-1);
      while (len--)
        *lastd-- = *lasts--;
    }
  return dest;
}
```
